spec/ics-2-consensus-verification/README.md

```diff
@@ -0,0 +1,190 @@
+
+* Blockchains, defined as an infinite list of `Header` starting from its genesis, is linear; no 
+conflicting `Header`s can be both validated, thus no data can be rewritten after it has been 
+committed. Two `Header`s are conflicting when both has same height but not equal.
+
+* Verifiers can verify future `Header`s using an existing `RootOfTrust`. When the verifier 
+validates it, the verified header is in the canonical blockchain.
+
+* `RootOfTrust`s contains an accumulator root(ICS23) that the other logics can verify whether 
+key-value pairs exists or not with it.
+
+### Technical Specification
+
+#### Definitions
+
+* `RootOfTrust` is a blockchain commit which contains an accumulator root and the requisite 
+  state to verify future roots, stored in one blockchain to verify the state of the other.
+  Defined as 2-tuple `(v :: Header -> (Error|RootOfTrust), r :: AccumulatorRoot)`, where
+    * `v` is the verifier, proves child `Header.p` and returns the updated `RootOfTrust`
+    * `r` is the `AccumuatorRoot`, used to prove internal state
+
+* `Header` is a blockchain header which provides information to update `RootOfTrust`, 
+  submitted to one blockchain to update the stored `RootOfTrust`.
+  Defined as 3-tuple `(p :: HeaderProof, v :: Maybe<Header -> (Error|RootOfTrust)>,
+  r :: AccumulatorRoot)`, where
+    * `p` is the commit proof used by `RootOfTrust.v` to verify
+    * `v` is the new verifier, if needed to be updated
+    * `r` is the new `AccumulatorRoot` which will replace the existing one
+ 
+* `Consensus` is a blockchain consensus algorithm which generates valid `Header`s.
+  Defined as a function `RootOfTrust -> [Transaction] -> Header`
+
+* `Blockchain` is a subset of `(Consensus, RootOfTrust, [Header])`, generated 
+by a `Consensus`.
+
+#### Requirements
+
+Consensus verification requires the following accumulator primitives with datastructures and
+properties as defined in ICS23:
+
+* `AccumulatorRoot`
+
+#### Subprotocols
+
+##### Verifier
+
+Verifiers prove new `Header` generated from a blockchain. It is expected to prove a `Header` 
+efficiently; more efficiently than replaying `Consensus` logic for given parent `Header` and the
+transactions. `Header.p` provides the proof that the verifier can use. Verifiers assume the
+following properties will be satisfied for the `Header`s:
+
+1. `Header`s have no more than one direct child
+ 
+* Satisfied if: deterministic safety
+* Possible violation scenario: validator double signing, deep chain reorg
+
+2. `Header`s have at least one direct child
+
+* Satisfied if: liveness, lightclient verifier continuability
+* Possible violation scenario: synchronized halt, incompatible hard fork
+
+3. `Header`s' accumulator root are valid transition from the parents'
+
+* Satisfied if: decentralized block generation, well implemented state machine
+* Possible violation scenario: invariant break, validator cartel
+
+// XXX: should it be described on the connection spec?
```

Comment:
I think we can move this to downstream specs. ICS 3 handles closing a connection on equivocation.

Can you make an issue to discuss the possible challenge period? That's a good idea but we should discuss it a bit first, not sure quite where to put it yet.